src/topology/metric_space/infsep.lean

```diff
@@ -0,0 +1,363 @@
+/-
+Copyright (c) 2022 Wrenna Robson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wrenna Robson
+-/
+import topology.metric_space.basic
+
+/-!
+
+Draft implementation of infimum separation.
+
+This is NOT A FINISHED FILE.
+In particular it needs: documentation, missing lemmas, and possibly other stuff.
+-/
+
+variables {α β : Type*}
+
+section extras
+
+lemma set.not_subsingleton_iff {s : set α} : ¬ s.subsingleton ↔ ∃ x y ∈ s, x ≠ y :=
+by simp_rw [set.subsingleton, not_forall]
+
+open_locale ennreal
+
+lemma le_edist_pi_iff {π : β → Type*} [fintype β] [nonempty β]
+  [Π b, pseudo_emetric_space (π b)] {f g : Π b, π b} {d : ℝ≥0∞} :
+  d ≤ edist f g ↔ ∃ b, d ≤ edist (f b) (g b) :=
+begin
+  by_cases hd : ⊥ < d,
+  { exact (finset.le_sup_iff hd).trans (by simp only [finset.mem_univ, exists_true_left]) },
+  { rw not_bot_lt_iff at hd, rw hd, simp only [ennreal.bot_eq_zero, zero_le, exists_const] }
+end
+
+end extras
+
+namespace set
+
+section infesep
+open_locale ennreal
+open function
+/-- The "extended infimum separation" of a set with an edist function. -/
+noncomputable def infesep [has_edist α] (s : set α) := ⨅ (x ∈ s) (y ∈ s) (hxy : x ≠ y), edist x y
+
+section has_edist
+variables [has_edist α] {x y : α} {s : set α}
+
+lemma le_infesep_iff {d} : d ≤ infesep s ↔ ∀ x y ∈ s, x ≠ y → d ≤ edist x y :=
+by simp_rw [infesep, le_infi_iff]
+
+lemma le_infesep_image_iff {d} {f : β → α} {s : set β} :
+  d ≤ infesep (f '' s) ↔ ∀ x y ∈ s, f x ≠ f y → d ≤ edist (f x) (f y) :=
+by simp_rw [le_infesep_iff, ball_image_iff]
+
+lemma le_infesep_image_injective_iff {d} {f : β → α} (hf : injective f) {s : set β} :
+  d ≤ infesep (f '' s) ↔ ∀ x y ∈ s, x ≠ y → d ≤ edist (f x) (f y) :=
+by simp_rw [le_infesep_image_iff, hf.ne_iff]
```

Comment:
This lemma feels weird to me; the name is strange, and it feels much more natural to use the two lemmas separately

Reply:
Yeah, if you feel this I'm fine with it: I was just worrying what the best generalisation of the diam equivalent was.